<img src="figs/cit_hub_logo.png" alt="logo" width="300"/>

# Introduction to Python

If you're diving into programming for the first time, learning Python fundamentals is the place to start. There's no limit to what you can do by learning Python programming, and expanding your skill set means expanding your career opportunities — and your earning potential.

This course requires no pre-requsitions, no previous math, stats, or programming experience required — so you can start learning right away. 

Here are a few takeaways you can expect by the end of this lesson:

- How computer programs work
- How to write and execute computer programs
- How to perform mathematical operations in Python
- How to use code comments
- How to save values using variables
- How to work with numbers and text data

During our journey, we will author several **portfolio projects**. In this Guided Projects we'll have some fun and analyze a lot of real-world datasets to extract meaningful information out of them.

To build the technologies we desire, we need to give the computer the proper instructions. When we give instructions to a computer, we say that we're **programming** it.

To **program a computer**, we need to write the instructions in a special language, which we call a *programming language*. In this lesson, we'll start learning Python, the most popular programming language.

## Python as a Calculator. Numbers and More

We can instruct the computer to add two numbers together, say `23 + 7`, and print the result. To do that, you'll need to do the following:

In [1]:
print(23 + 7)

30


However, we can ask the computer to perform more than just one calculation. Let's demonstrate this using data from a real dataset. Throughout this lesson, we'll learn core Python programming concepts in the context of real data. We'll be using data from the table below, which provides some information about five mobile applications from the iOS store:

![example dataset](figs/1-app_dataset.png)

In the examples below we instruct the computer to perform two calculations:

- The cost of purchasing the `Facebook` app `(0.00)` and the `Fruit Ninja` Classic app `(1.99)`
- The difference in the total number of `user ratings` submitted for `Facebook (2974676)` and `Instagram (2161558)`

In [2]:
print(0.00 + 1.99)
print(2974676 - 2161558)

1.99
813118


To display both results, we need to use the **`print()` function**. The total cost of purchasing Facebook and Fruit Ninja Classic is `1.99`. Facebook has `813118` more user rating results than Instagram.

## Syntax Rules - Grammar is Everywhere

Previously, we sent the computer two instructions and wrote each on a separate line. If we were to put them both on the same line, we'd get an error:

In [3]:
print(0.00 + 6.99) print(4.5 - 3.5)

SyntaxError: invalid syntax (<ipython-input-3-68f08bce7f28>, line 1)

`print(0.00 + 6.99) print(4.5 - 3.5)` resulted in text describing a **syntax** error. This is because all programming languages — Python included — have syntax rules. Each line of instructions has to comply with these rules.
Running into errors is common when we're programming, and we'll constantly learn about errors and how to fix them. 

Collectively, we call all these *instructions* we write **code**. We call each line of instruction a line of code.

When we write code, we *program* the computer to do something. For this reason, we also call the code we write a **computer program (or a program)**.

The code we write serves as input to the computer. We call the result of executing the code output.

![computer program](figs/2-code.svg)


### Comments - a Necessary Break when  Coding
The computer executes code from the first line downward and ignores blank lines. Besides blank lines, the computer also ignores any sequence of characters that comes to the right of the `#` symbol. The sequence of characters that follows the `#` symbol is a **code comment**. We can use code comments to add information about our code:

In [6]:
print(1.99)   # App cost for Fruit Ninja Classic
print(698516) # Rating count
print(4.5)    # Average user rating

1.99
698516
4.5


In Python, comments serve multiple purposes

- Comments can be used to explain Python code.

- Comments can be used to make the code more readable.

- Comments can be used to prevent execution when testing code.

In [7]:
# comments can be in one line
print("Nice commenting!")

Nice commenting!


In [8]:
# or
# separated in multiple lines
# for improved readability
print("I see clearer now")

I see clearer now


Or, you can use a multiline string.

Since Python will ignore string literals that are not assigned to a variable, you can add a multiline string (triple quotes) in your code, and place your comment inside it:

In [9]:
"""
This is a comment
written in
more than just one line
"""
print("Hi, there!")

Hi, there!


## Numbers in Python

Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python

### Basic arithmetic


In [10]:
# Addition
2+1

3

In [11]:
# Subtraction
2-1

1

In [12]:
# Multiplication
2*2

4

In [13]:
# Division
3/2

1.5

In [14]:
# Floor Division
7//4

1

In [15]:
# Modulo / remainder
7%4

3

In [16]:
# Power
2**3

8

In [17]:
# can also work like this
4**0.5

2.0

In [18]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [19]:
# Can use parentheses to specify order
(2+10) * (10+3)

156

## Programming Python Variables

We'll learn how to save values in Python, how to update values, and how to use syntax shortcuts to update values. We use a single equals sign (`=`) to assign labels to variables. Let's see a few examples of how we can do this.

In [20]:
# Let's create a variable called "result" and assign it the number 3.98
result = 3.98

In [21]:
# Adding the objects where the variable is stored
a+a

NameError: name 'a' is not defined

What happens on reassignment? Will Python let us write it over?

In [22]:
# Reassignment
a = 15
# Check
a

15

When we run the code `result = 3.98`, the value `3.98` gets saved in the computer memory. The computer memory has many storage locations, and `3.98` gets saved to one particular location.

![variables](figs/1-variables.svg)

The storage location for `3.98` is a **variable**. When we ran the code `result = 3.98`, we stored `3.98` in a variable (storage location) named `result`, so `result` is a variable name.

Using variable names can be a very useful way to keep track of different variables in Python. For example:

In [23]:
# Use object names to keep better track of what's going on in your code!
my_income = 100

tax_rate = 0.1

my_taxes = my_income*tax_rate

In [24]:
# Show my taxes!
my_taxes

10.0

These are the two syntax rules we need to follow when we're naming variables:

1. We must use only letters, numbers, or underscores (we can't use apostrophes, hyphens, spaces, etc.).
2. Variable names can't begin with a number.

![variables](figs/3-variables.svg)

Note that variable names are case sensitive, which means that a variable named `result` is different than a variable named `Result`. **It is common practice to write python variables in all lowercase letters** such as `taxes = 10.5`, and in case of multiple words to seperate them with the `_` symbol, such as `my_taxes = 10.5`

We can update the value stored in a variable. Below, we first store `1.99` in the variable `app_costs`, and then we update `app_costs` to store `6.99` instead.

In [25]:
app_costs = 1.99
print(app_costs)

app_costs = 6.99
print(app_costs)

1.99
6.99


We can also update a variable doing arithmetic operations. There are several syntax shortcuts we can use to update a variable when we're doing arithmetic operations. Below is a table with some syntax shortcuts we can use:

![variables](figs/4-variables_shortcut.svg)

Note that we can only use these operators `(+=, -=, *=, /=, **=)` to update a variable. This means the variable we're updating must already store a value. In other words, the variable must already be defined. When we try to update a variable that we haven't defined, we get an error called `NameError`.

In [26]:
# Error:
app_cost += 6.99

NameError: name 'app_cost' is not defined

In Python, the `=` operator tells us that the value on the right is assigned to the variable on the left. It doesn't tell us anything about equality. We call `=` an assignment operator, and we read code like `x = 5` as "five is assigned to x" or "x is assigned five," not "x equals five."

### Types of Variables in Python

In computer programming, we classify values into different **types** — or **data types**. The type of value offers the computer the required information about how to process that value. Depending on the type, the computer will know how to store a value in memory, or what operations we can and can't perform on a value.

We can check what type of object is assigned to a variable using Python's built-in `type()` function. Common data types include:

- `int` (for integer, i.e., natural positive and negative numbes, e.g., 2)
- `float` (for floating point / decimal numbers, e.g., 3.14)
- `str` (for string, e.g, a given text such as "hello")
- `bool` (for Boolean True/False)
- `list` (for a list of elements)
- `tuple` (for a tuple of elements)
- `dict` (for dictionary)
- `set` (for a unique set of elements)

We have already worked with numbers. Let us work with strings. 

## Strings in Python

Strings are used in Python to record text information, such as names. Strings in Python are actually a sequence, which basically means Python keeps track of every element in the string as a sequence. For example, Python understands the string "hello' to be a sequence of letters in a specific order. This means we will be able to use indexing to grab particular letters (like the first letter, or the last letter).

This idea of a sequence is an important one in Python and we will touch upon it later on in the future.

### Creating a string

To create a string in Python you need to use either single quotes or double quotes. For example:

In [27]:
# Single word
'hi'

'hi'

In [28]:
# Entire phrase 
'Hi there. Python is cool'

'Hi there. Python is cool'

In [29]:
# We can also use double quote
"Yes, Python is even cooler now"

'Yes, Python is even cooler now'

In [30]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: invalid syntax (<ipython-input-30-da9a34b3dc31>, line 2)

The reason for the error above is because the single quote in I'm stopped the string. You can use combinations of double and single quotes to get the complete statement.



In [31]:
"Now that is what I'm talking about"

"Now that is what I'm talking about"

### Printing a string
Using Jupyter notebook with just a string in a cell will automatically output strings, but the correct way to display strings in your output is by using a print function.

In [32]:
# We can simply declare a string
'Hello World'

'Hello World'

In [33]:
# Note that we can't output multiple strings this way
'Hello World 1'
'Hello World 2'

'Hello World 2'

In [34]:
# We can use a print statement to print a string

print('Hello World 1')
print('Hello World 2')
print('Use \n to print a new line')
print('\n')
print('See what I mean?')

Hello World 1
Hello World 2
Use 
 to print a new line


See what I mean?


### String fundamentals
We can also use a function called `len()` to check the length of a string! Note that it counts all of the characters in the string, including spaces and punctuation.

In [35]:
len('A great sentence!')

17

### String indexing

We know strings are a sequence, which means Python can use indexes to call parts of the sequence. Let's learn how this works.

In Python, we use brackets `[]` after an object to call its index. We should also note that indexing starts at `0` for Python. Let's create a new object called s and then walk through a few examples of indexing.

In [36]:
# Assign s as a string
s = 'Hi there!'

In [37]:
#Check
s

'Hi there!'

In [38]:
# Show first element (in this case a letter)
s[0]

'H'

In [39]:
s[3]

't'

We can use a : to perform slicing which grabs everything up to a designated point. For example:

In [40]:
# Grab everything past the first term all the way to the length of s which is len(s)
s[3:]

'there!'

In [41]:
# Grab everything UP TO the 3rd index
s[:3]

'Hi '

In [42]:
#Everything
s[:]

'Hi there!'

We can also use negative indexing to go backwards.

In [43]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'!'

In [44]:
# Grab everything but the last letter
s[:-1]

'Hi there'

We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:

In [45]:
# Grab everything, but go in steps size of 1
s[::1]

'Hi there!'

In [46]:
# Grab everything, but go in step sizes of 2
s[::2]

'H hr!'

In [47]:
# We can use this to print a string backwards
s[::-1]

'!ereht iH'

### String properties

It's important to note that strings have an important property known as immutability. This means that once a string is created, the elements within it can not be changed or replaced. For example:

In [48]:
s

'Hi there!'

In [49]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Something we can do is concatenate strings!

In [50]:
s

'Hi there!'

In [51]:
# Concatenate strings!
s + ' Python is cool!'

'Hi there! Python is cool!'

In [52]:
# We can reassign s completely though!
s = s + ' Python is cool!'
s

'Hi there! Python is cool!'

We can use the multiplication symbol to create repetition!

In [53]:
sentence = "Python is cool! "

In [54]:
sentence * 5

'Python is cool! Python is cool! Python is cool! Python is cool! Python is cool! '

### String methods

Objects in Python usually have built-in methods. These methods are functions inside the object (we will learn about these in much more depth later) that can perform actions or commands on the object itself.

We call methods with a period and then the method name. Methods are in the form:

object.method(parameters)

Where parameters are extra arguments we can pass into the method. Don't worry if the details don't make 100% sense right now. Later on we will be creating our own objects and functions!

Here are some examples of built-in methods in strings:

In [55]:
s

'Hi there! Python is cool!'

In [56]:
# Upper Case a string
s.upper()

'HI THERE! PYTHON IS COOL!'

In [57]:
# Lower case
s.lower()

'hi there! python is cool!'

In [58]:
# Split a string by blank space (this is the default)
s.split()

['Hi', 'there!', 'Python', 'is', 'cool!']

### String formatting

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

In [59]:
player = 'Thomas'
points = 33

'Last night, '+player+' scored '+str(points)+' points.'  # concatenation

f'Last night, {player} scored {points} points.'          # string formatting

'Last night, Thomas scored 33 points.'

There are three ways to perform string formatting.

- The oldest method involves placeholders using the modulo `%` character.
- An improved technique uses the `.format()` string method.
- The newest method, introduced with Python 3.6, uses formatted string literals, called `f-strings`.

##### Formatting with the `.format()` method
The syntax is:

In [60]:
'String here {} then also {}'.format('something1','something2')

'String here something1 then also something2'

In [61]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


In [62]:
# Inserted objects can be assigned keywords:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))


First Object: 1, Second Object: Two, Third Object: 12.3


In [63]:
# Inserted objects can be reused, avoiding duplication:
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.


##### Formatted String Literals (`f-strings`)
Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. 

In [64]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


For more information, visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

## Summary

Congratulations on completing this lesson! Let's take a moment to review what we learned in this lesson:

- How computer programs work
- How to write and execute computer programs
- How to use code comments
- How to perform arithmetical operations in Python

And also...
- How to save our data using variables
- How to update variables
- How to use syntax shortcuts to update variables
- How to work with integers and floats
- How to work with text data using strings